pkg/pillar/cmd/zedrouter/flowstats.go

```diff
@@ -0,0 +1,738 @@
+	pub := ctx.pubAppNetworkStatus
+	items := pub.GetAll()
+	for _, st := range items {
+		status := cast.CastAppNetworkStatus(st)
+		for i, ulStatus := range status.UnderlayNetworkList {
+			log.Infof("===FlowStats: (index %d) AppNum %d, VifInfo %v, IP addr %v, Hostname %s\n",
+				i, status.AppNum, ulStatus.VifInfo, ulStatus.AllocatedIPAddr, ulStatus.HostName)
+
+			// build an App-IPaddress cache indexed by App-number
+			if ulStatus.AllocatedIPAddr != "" {
+				tmpIPs := net.ParseIP(ulStatus.AllocatedIPAddr)
+				instData.appIPAddrs[status.AppNum] = append(instData.appIPAddrs[status.AppNum], tmpIPs)
+			}
+
+			// Fill in the bnNet indexed by bridge-name, used for loop through bridges, and Scope
+			ulconfig := ulStatus.UnderlayNetworkConfig
+			intfAttr := aclAttr{
+				vif:     ulStatus.VifInfo.Vif,
+				bridge:  ulStatus.Bridge,
+				netUUID: ulconfig.Network,
+			}
+			instData.bnNet[ulStatus.Bridge] = intfAttr
+
+			// build an App list cache, used for loop through all the Apps
+			if instData.appNet[status.AppNum] == nilUUID {
+				instData.appNet[status.AppNum] = status.UUIDandVersion.UUID
+				log.Infof("===FlowStats: appNet appNum %d, uuid %v\n", status.AppNum, instData.appNet[status.AppNum])
+			}
+
+			// build an acl cache indexed by app/aclnum, from flow MARK, we can get this aclAttr info
+			tmpMap := instData.ipaclattr[status.AppNum]
+			if tmpMap == nil {
+				tmpMap := make(map[int]aclAttr)
+				instData.ipaclattr[status.AppNum] = tmpMap
+			}
+			for _, rule := range ulStatus.ACLRules {
+				if rule.IsUserConfigured == false { // only include user defined rules
+					continue
+				}
+
+				var tempAttr aclAttr
+				tempAttr.aclNum = uint32(rule.RuleID)
+				tempAttr.chainName = rule.ActionChainName
+				tempAttr.aclName = rule.RuleName
+				tempAttr.tableName = rule.Table
+				tempAttr.vif = ulStatus.VifInfo.Vif
+				tempAttr.bridge = ulStatus.Bridge
+				if len(rule.Action) >= 2 { // '-j ACCEPT', '-j drop-all-bn1-nbu2x1'
+					tempAttr.action = rule.Action[1]
+				}
+				if _, ok := instData.ipaclattr[status.AppNum][int(rule.RuleID)]; !ok { // fake j as the aclNUM
+					instData.ipaclattr[status.AppNum][int(rule.RuleID)] = tempAttr
+				} else {
+					preAttr := instData.ipaclattr[status.AppNum][int(rule.RuleID)]
+					// the the entry exist, and already has the aclNum and bridge name, skip
+					if preAttr.aclNum == 0 || preAttr.bridge == "" {
+						instData.ipaclattr[status.AppNum][int(rule.RuleID)] = tempAttr
+					}
+				}
+			}
+		}
+	}
+}
+
+// DNSMonitor : DNS Query and Reply monitor on bridges
+func DNSMonitor(bn string, bnNum int) {
+	var (
+		err error
+		//action      string
+		snapshotLen int32 = 1280             // draft-madi-dnsop-udp4dns-00
+		promiscuous       = true             // mainly for switched network
+		timeout           = 10 * time.Second // collect enough packets in 10sec before processing
+		handle      *pcap.Handle
+		filter      = "udp and port 53"
```

Comment:
Need to also look for tcp port 53 since DNS can fall back to TCP for large responses.

Reply:
yes, will add tcp into this.

Reply:
actually this is more tricky than I thought. apparently the goland packet somehow does not get the TCP port 53 correctly. I'll make a comment to this, and come back to do more research on how to implement this for TCP DNS records.